### ISSUE TITLE
Add interfaces accepting scorpio data structures rather than a list of builtin types

### ISSUE THREAD

**Issue by bartgol** [E3SM-Project/scorpio#314]
I was wondering if there is a plan to add a set of interfaces accepting scorpio data structures instead of a bunch of raw types (int, double, pointers,...).

For instance, this function:

```
int PIOc_createfile(int iosysid, int *ncidp,  int *iotype, const char *fname, int mode);
```

could be made more user friendly (for a C/C++ downstream user, that is) like the following (I'm using same name, but C does not support overload, so might have to be adjusted):

```
int PIOc_createfile(file_desc_t* file_desc);
```
This function can unpack needed input from desc (mode, iosystem, iotype), also checking that they are valid, and stuff the output (ncidp) in side it.

The reason is that the C/C++ user might have a cleaner code if he/she already stores PIO stuff using PIO structures, and finding a line

```
PIOc_createfile(my_file.iosystem->iosysid, &my_file.pio_ncid,&my_file.iotype,my_file.fname,my_file.mode);
```
might be harder to parse (and a waste of keystrokes ;-) ) than
```
PIOc_createfile(my_file);
```

**Comment by jayeshkrishna** [comment]
Hmm, I am not sure if we want to change the C API right now.
However we can start thinking about new C++ interfaces for Scorpio that might be more intuitive to the user (Instead of filling up a structure with Scorpio-specific information add new C++ classes that is more intuitive to the C++ developer).

**Comment by jayeshkrishna** [comment]
Also (just as a word of caution) some of the internal data structures have been exposed in pio.h (the Scorpio user header file) right now, but we will be moving away from that in future.

**Comment by bartgol** [comment]
> Hmm, I am not sure if we want to change the C API right now.

To be clear, I am not advocating for removing the current interfaces. You do need them in order for the flib to easily call the clib.

What I'm wondering is whether the scorpio team is thinking on a lighter interface for users who interface directly with the pioc lib.

**Comment by bartgol** [comment]
> Also (just as a word of caution) some of the internal data structures have been exposed in pio.h (the Scorpio user header file) right now, but we will be moving away from that in future.

Interesting. Might be worth adding a comment in it, or move those types in a `pio_internal_types.h` header (included by pio.h), before anyone decides to include the header and start using those types...

**Comment by jayeshkrishna** [comment]
Yeah, the plan is to move these types into an internal header file
